plugins/newsdownloader.koplugin/epubdownloadbackend.lua

```diff
@@ -1,24 +1,100 @@
+local NewsHelpers = require("http_utilities")
 local Version = require("version")
-local ffiutil = require("ffi/util")
-local http = require("socket.http")
 local logger = require("logger")
-local ltn12 = require("ltn12")
-local socket = require("socket")
 local socket_url = require("socket.url")
-local socketutil = require("socketutil")
 local _ = require("gettext")
-local T = ffiutil.template
 
-local EpubDownloadBackend = {
+local EpubBuilder = {
    -- Can be set so HTTP requests will be done under Trapper and
    -- be interruptible
    trap_widget = nil,
    -- For actions done with Trapper:dismissable methods, we may throw
    -- and error() with this code. We make the value of this error
    -- accessible here so that caller can know it's a user dismiss.
    dismissed_error_code = "Interrupted by user",
+   title = nil,
+   ncx_toc = nil,
+   ncx_manifest = nil,
+   ncx_contents = nil,
+   ncx_images = nil,
 }
-local max_redirects = 5; --prevent infinite redirects
+
+function EpubBuilder:new(o)
+    o = o or {}
+    self.__index = self
+    setmetatable(o, self)
+
+    return o
+end
+
+function EpubBuilder:build(abs_output_path)
+    -- Open the zip file (with .tmp for now, as crengine may still
+    -- have a handle to the final epub_path, and we don't want to
+    -- delete a good one if we fail/cancel later)
+    local tmp_path = abs_output_path .. ".tmp"
+    local ZipWriter = require("ffi/zipwriter")
+    local epub = ZipWriter:new{}
+
+    if not epub:open(tmp_path) then
+        logger.dbg("Failed to open tmp_path")
+        return false
+    end
+
+    epub:add("mimetype", "application/epub+zip")
+    epub:add("META-INF/container.xml", [[
+<?xml version="1.0"?>
+<container version="1.0" xmlns="urn:oasis:names:tc:opendocument:xmlns:container">
+  <rootfiles>
+    <rootfile full-path="OEBPS/content.opf" media-type="application/oebps-package+xml"/>
+  </rootfiles>
+</container>]])
+
+    -- Add the manifest.
+    if not self.ncx_manifest or #self.ncx_manifest == 0 then
+        error(_("EPUB does not contain a valid manifest."))
```

Comment:
Are these errors intended to be translated?

Reply:
Yes... The sync process will deliver that error to the user if it occurs

Reply:
How could something you're adding yourself ever not be there?

Reply:
(Forgot to submit comments… oops.)

Reply:
> How could something you're adding yourself ever not be there?

Yeah, that's probably overkill having the error. The only instance where the manifest wouldn't be set yet is if some programmer doesn't call the function to add the manifest.

Reply:
Developer-only errors are good, but shouldn't be shown to the user or translated. (Shouldn't in the sense that we don't want to waste the translator's valuable time on that. ;-) )